File: Algorithm/C++/988. Smallest String Starting From Leaf.cc
/*
 * Created on Sun Feb 03 2019 18:31:43
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

// recursive
// std::min can not compare empty string

#include <string>

using std::string;

class Solution {
 public:
  string smallestFromLeaf(TreeNode* root) {
    if (root == nullptr) return "";
    string left = smallestFromLeaf(root->left);
    string right = smallestFromLeaf(root->right);
    if (left.length() == 0) {
      return right + char(root->val + 97);
    } else if (right.length() == 0) {
      return left + char(root->val + 97);
    } else {
      return std::min(left, right) + char(root->val + 97);
    }
  }
};

File: Algorithm/Go/988. Smallest String Starting From Leaf.go
/*
 * Created on Sun Feb 03 2019 18:41:25
 * Author: WuLC
 * EMail: liangchaowu5@gmail.com
 */

 /**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

// recursive
import "strings"

func smallestFromLeaf(root *TreeNode) string {
    if root == nil {
		return ""
	}
	left, right := smallestFromLeaf(root.Left), smallestFromLeaf(root.Right)
	if len(left) == 0 {
		return right + string(root.Val + 97)
	} else if len(right) == 0  {
		return left + string(root.Val + 97)
	} else {
		return min(left, right) + string(root.Val + 97) 
	}
}

func min(s1, s2 string) string {
	if strings.Compare(s1, s2) < 0 {
		return s1
	} else {
		return s2
	}
}